File: app/src/main/java/com/example/tah/ui/habit/HabitRecyclerViewAdapter.java
package com.example.tah.ui.habit;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.example.tah.R;
import com.example.tah.models.Habit;
import com.example.tah.ui.main.AddAndDetailsActivity;
import com.example.tah.utilities.ViewHabitTime;

import java.util.List;
import java.util.Locale;
import java.util.Map;

public class HabitRecyclerViewAdapter
        extends RecyclerView.Adapter<HabitRecyclerViewAdapter.ViewHolder>
        implements ViewHabitTime{

    private final List<Habit> habits;
    private final Context context;

    public HabitRecyclerViewAdapter(Context context, List<Habit> items) {
        this.context = context;
        habits = items;
    }

    public void update(List<Habit> habits){
        this.habits.clear();
        this.habits.addAll(habits);
        this.notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_habit, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.bind(habits.get(position));
    }

    @Override
    public int getItemCount() {
        return habits.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        final View mView;
        final TextView name;
        final TextView description;
        final TextView time;


        public Habit habit;


        public ViewHolder(View view) {
            super(view);
            mView = view;
            name = view.findViewById(R.id.item_name);
            description = view.findViewById(R.id.description);
            time = view.findViewById(R.id.time);
        }

        private void bind(Habit habit){
            this.habit = habit;

            Map<String, String> timeMap = getTimeStrings(habit.getSessionLength());

            String timeText = String.format(Locale.ENGLISH,
                    "Time left: %s:%s:%s",
                    timeMap.get("Hours"),
                    timeMap.get("Minutes"),
                    timeMap.get("Seconds"));

            name.setText(habit.getName());
            description.setText(habit.getDescription());
            time.setText(timeText);

            setOnClickListeners();
        }

        private void setOnClickListeners(){
            itemView.setOnClickListener(v -> {
                Intent intent = new Intent(context, AddAndDetailsActivity.class);
                intent.putExtra("fragmentId", Habit.Companion.getDetailsView());
                intent.putExtra("habitId", habit.getId());
                context.startActivity(intent);
            });
        }
    }
}

File: app/src/main/java/com/example/tah/ui/habit/HabitRecyclerViewAdapter.kt
package com.example.tah.ui.habit

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.tah.databinding.ItemHabitBinding
import com.example.tah.models.Habit
import com.example.tah.ui.main.AddAndDetailsActivity
import com.example.tah.utilities.ViewHabitTime
import java.util.*

class HabitRecyclerViewAdapter(
    private val context: Context,
    private val habits: MutableList<Habit>)
: RecyclerView.Adapter<HabitRecyclerViewAdapter.ViewHolder>(),
    ViewHabitTime
{

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemHabitBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(habits[position])

    override fun getItemCount(): Int = habits.size

    fun update(items: List<Habit>){
        habits.clear()
        habits.addAll(items)
        notifyDataSetChanged()
    }

    inner class ViewHolder(
        private val binding: ItemHabitBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        private lateinit var habit: Habit

        fun bind(habit: Habit){
            this.habit = habit

            val timeMap = getTimeStrings(habit.sessionLength)
            val timeText = String.format(Locale.ENGLISH,
                    "Time left: %s:%s:%s",
                    timeMap["Hours"],
                    timeMap["Minutes"],
                    timeMap["Seconds"])

            with(binding){
                itemName.text = habit.name
                time.text = timeText

                if (habit.description.isNullOrEmpty()){
                    description.visibility = View.GONE
                } else {
                    description.visibility = View.VISIBLE
                    description.text = habit.description
                }
            }
            setOnClickListeners()
        }

        private fun setOnClickListeners(){
            itemView.setOnClickListener {
                val intent = Intent(context, AddAndDetailsActivity::class.java)
                intent.putExtra("fragmentId", Habit.getDetailsView())
                intent.putExtra("habitId", habit.id)
                context.startActivity(intent)
            }
        }
    }
}